### PR TITLE
Document metrics generated within OpenWhisk

### PR DESCRIPTION
**Pull request by chetanmeh**
OpenWhisk generates quite a few metrics via [Kamon](http://kamon.io). This PR aims to document all currently generated metrics such that system administrators can configure the right metrics for monitoring

See [here](https://github.com/chetanmeh/incubator-openwhisk/blob/metrics-docs/docs/metrics.md#metric-details) for the rendered HTML version of doc

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by rabbah** [comment]
Fantastic!

**Comment by rabbah** [comment]
@vvraskin @mhenke1 would you review this? 

@chetanmeh this is very useful thanks for doing this.

File: docs/metrics.md [hunk 1]
```diff
@@ -75,6 +75,183 @@ The docker image exposes StatsD via the (standard) port 8125 and a Grafana dashb
...
+  * `openwhisk.counter.invoker_docker.unpause_start`
+  * `openwhisk.counter.invoker_docker.unpause_error`
+  * `openwhisk.histogram.invoker_docker.unpause_finish`
+  * `openwhisk.histogram.invoker_docker.unpause_error`
+
+#### Kafka Metrics
+
+Metrics below are emitted per kafka topic.
+
+* `openwhisk.histogram.kafka_<topic name>.delay_start` - Time delay between when a message was pushed to kafka and when it is read within a consumer.
```
Comment (vvraskin) [inline docs/metrics.md]: should we also mention that Delay is being emitted for each pool by Invoker, while Queue metric is emitted every 10 seconds?

File: docs/metrics.md [hunk 2]
```diff
@@ -75,6 +75,183 @@ The docker image exposes StatsD via the (standard) port 8125 and a Grafana dashb
...
+
+Below are some of the important metrics emitted by OpenWhisk setup
+
+#### Controller metrics
+
+Metrics below are emitted from with a Controller instance.
+
+##### Controller Startup
+
+* `openwhisk.counter.controller_startup<controller_id>_count` (counter)
```
[the remark this answers is not in the document]
Comment (chetanmeh) [reply docs/metrics.md]: Checked the stats in our setup and they are named like `openwhisk.counter.controller_startup0_count`. Also in code the prefix is `startup$id`

https://github.com/apache/incubator-openwhisk/blob/8b5abe7bbeb4464af586c1993fc3590e0fe8d516/common/scala/src/main/scala/whisk/common/Logging.scala#L257

File: docs/metrics.md (hunk 1 above)
Comment (chetanmeh) [reply docs/metrics.md]: Missed that it was be done via a scheduled task. Per default config it should be emitted every 60 seconds

**Comment by vvraskin** [comment]
@chetanmeh 
Which backend do you use for the system metrics?
If you use Grafana+InfluxDB, it allows us to aggregate metrics of several components using regex filters. I'm wondering whether the same approach would be feasible...

**Comment by chetanmeh** [comment]
@vvraskin  We use Datadog and it [does not seem][2] to support [regex in metric names][1]. Which makes it tricky to aggregate metrics across various invokers. 

Does Grafana+Influx support tags? Looking for common denominator here

[1]: https://docs.datadoghq.com/graphing/faq/when-i-query-can-i-use-wildcards-in-metric-names-and-events/
[2]: https://trello.com/c/WtJXrGQj/18-datadog-wildcard-regex-matches-for-metric-names-1

**Comment by vvraskin** [comment]
Our statsd integration with influx won't support tagging, but I think that we already have tagging option for Kamon introduced in this PR https://github.com/apache/incubator-openwhisk/commit/6f1a445f22f827bd91f75b01bc82988054febbc8

**Comment by chetanmeh** [comment]
Opened #3917 to discuss possible options for supporting setups which do not support regex

**Pull request merged**